### PR TITLE
feature: Basic auth support with fission router

### LINKED ISSUE #2277: Basic auth support with fission router
**Is your feature request related to a problem? Please describe.**

Currently, Fission doesn't have any inbuilt support for authentication at the Fission router.
The fission router is the entry point for all function calls made by the Fission user.

**Describe the solution you'd like**
Fission should have basic auth support which can be enabled while deploying function.
It could be getting a token with standard basic auth and using it for all requests made fission function calls.

**Describe alternatives you've considered**
In the current scenario, the user has to configure basic auth by himself at the ingress level, which does provide a layer of authentication for external calls. But if a user is using the Kubernetes service only then he cant configure authentication for all fission function calls.

**Additional context**

Probable set of changes required:
- Decide standard auth method strategy to be used 
- Generate random credentials at the time of deploying Fission
- Configure auth middleware in Fission router(changes required) so that all requests are checked if they are authenticated then forwarded to respective function handlers
- We would also need to ensure necessary changes are made in Fission CLI

### PR DESCRIPTION
**Pull request by praddy26**
## Description

This PR adds new feature to enable authentication for fission routers. 
This is **a optional feature** which can be enabled/disabled as per the needs of the user. 
On enabling this feature, an endpoint for authentication will be registered in the router and all function API calls with be routed via authentication middleware. 
The user has to first create auth token by passing credentials. 
The token generated has to be then passed in the authorization header for all susbequent API calls to access the function.

This feature can be turned on by setting the key `authentication.enabled` in `charts/fission-all/values.yaml ` to `true` during the time of deployment. 
Along with this there are other keys like 
- `authUrlPath`
-  `authUsername`
- `jwtSigningKey`
- `jwtExpiryTime` 
- `jwtIssuer` 

The description for these parameters are provided in the same file as comments.

On enabling this feature a secret by name `router` will be created in the `fission` namespace with `username`, randomly generated `password` and `jwtSigningKey`. 
This secret will be mounted as volume on the router pod.
Also the other values mentioned above will be added to the existing `feature-config` configmap and will be mounted as volume on the router pod.

fission function test command:
---------------------------------
-  If default `authUrlPath` is not used, then the custom path has to be exported to the environment variable `FISSION_AUTH_URI`

  ```bash
  export FISSION_USERNAME=$(kubectl get secrets/router --template={{.data.username}} -n fission | base64 -d)
  export FISSION_PASSWORD=$(kubectl get secrets/router --template={{.data.password}} -n fission | base64 -d)
  export FISSION_AUTH_TOKEN=$(fi token create --username $FISSION_USERNAME --password $FISSION_PASSWORD)
  ```

-  For testing function token is picked from the env variable `FISSION_AUTH_TOKEN`, 
  
  ```bash
  fission function --name <functionName>
  ```

-  If env variable is not set then token has to be passed in the header 

  ```bash
  fission function test --name <functionName> --header "Authorization:Bearer <token>" 
  ```

fission function API:
----------------------------------------------
-  Router port has to be port-forwarded to one of the port. ex: 3000
  
  ```bash
  export FISSION_USERNAME=$(kubectl get secrets/router --template={{.data.username}} -n fission | base64 -d)
  export FISSION_PASSWORD=$(kubectl get secrets/router --template={{.data.password}} -n fission | base64 -d)
  curl -v http://localhost:3000/auth/login -d '{"username":"'$FISSION_USERNAME'", "password":"'$FISSION_PASSWORD'"}'
  export FISSION_AUTH_TOKEN=<token>
  curl http://localhost:3000/<path> -H "Authorization: Bearer ${FISSION_AUTH_TOKEN}"
  ```

## Which issue(s) this PR fixes:

Fixes #2277

## Checklist:

- [x] I ran tests as well as code linting locally to verify my changes. 
- [x] I have done manual verification of my changes, changes working as expected.
- [ ] I have added new tests to cover my changes.
- [x] My changes follow contributing guidelines of Fission.
- [x] I have signed all of my commits.

### CODE REVIEW

File: charts/fission-all/templates/router/deployment.yaml [hunk 1]
```diff
@@ -35,6 +35,23 @@ spec:
         command: ["/fission-bundle"]
         args: ["--routerPort", "8888", "--executorUrl", "http://executor.{{ .Release.Namespace }}"]
         env:
+        {{- if .Values.authentication.enabled }}
+        - name: ROUTER_USERNAME
```
Comment (sanketsudake) [inline charts/fission-all/templates/router/deployment.yaml:39]: Should we prefer `AUTH_USERNAME` and `AUTH_PASSWORD` ?

File: charts/fission-all/templates/_helpers.tpl [hunk 2]
```diff
@@ -31,6 +31,13 @@ canary:
   prometheusSvc: {{ .Values.prometheus.serviceEndpoint | default "" | quote }}
   {{- end }}
   {{- printf "\n" -}}
+auth:
+  enabled: {{ .Values.authentication.enabled | default false }}
+  {{- if .Values.authentication.enabled }}
+  authUrlPath: {{ .Values.authentication.authUrlPath | default "/auth/login" | quote}}
+  jwtExpiryTime: {{ .Values.authentication.jwtExpiryTime | default 120 }}
```
Comment (sanketsudake) [inline charts/fission-all/templates/_helpers.tpl:38]: I dont see any limitation of this config as of now. But in future if we wish to support different auth methods we need to make sure auth config is flexible for that. Please see from your perspective also we need any changes here from that perspective.

File: charts/fission-all/templates/router/secret.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.authentication.enabled }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: router
+  labels:
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+type: Opaque
```
Comment (sanketsudake) [inline charts/fission-all/templates/router/secret.yaml:8]: Should we skip type here? This way user default secret method would be used while installation ?

File: charts/fission-all/templates/router/secret.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.authentication.enabled }}
+apiVersion: v1
```
Comment (sanketsudake) [inline charts/fission-all/templates/router/secret.yaml:2]: We can use `pre-install` hook for this resource ?
https://helm.sh/docs/topics/charts_hooks/

File: charts/fission-all/values.yaml [hunk 5]
```diff
@@ -479,6 +478,36 @@ prometheus:
...
+authentication:
+  ## set this flag to true if you need authorization
+  ## for all route calls
+  ##
+  enabled: false
+  ## authUrlPath is the router path for
+  ## authentication endpoint, defaults to /auth/login
+  ## deafult '/auth/login'
+  ##
+  authUrlPath: /auth/login
```
Comment (sanketsudake) [inline charts/fission-all/values.yaml:491]: Just wondering if tomorrow if we consider an external auth provider? Would it work with these params?

File: pkg/fission-cli/cmd/token/create.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	jsonValue, _ := json.Marshal(values)
+
+	kubeContext := input.String(flagkey.KubeContext)
+	// Portforward to the fission router
+	localRouterPort, err := util.SetupPortForward(util.GetFissionNamespace(), "application=fission-router", kubeContext)
+	if err != nil {
+		return err
+	}
+
+	authAuthenticatorUrl := "http://127.0.0.1:" + localRouterPort + "/auth/login"
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/token/create.go:72]: As URL is configurable, we can support some ENV variables. If the user sets the ENV variable, we would consider provided URL.

**Review by sanketsudake — CHANGES_REQUESTED** [review]

File: charts/fission-all/templates/router/deployment.yaml (hunk 1 above)
Comment (praddy26) [reply charts/fission-all/templates/router/deployment.yaml:39]: Fixed

File: charts/fission-all/templates/router/secret.yaml (hunk 3 above)
Comment (praddy26) [reply charts/fission-all/templates/router/secret.yaml]: Fixed

File: charts/fission-all/values.yaml (hunk 5 above)
Comment (praddy26) [reply charts/fission-all/values.yaml]: The auth URI  value is looked for in the env variable `FISSION_AUTH_URI`. If it is not set or if it is a empty string then it defaults to `/auth/login`

File: charts/fission-all/values.yaml [hunk 7]
```diff
@@ -483,6 +482,37 @@ prometheus:
...
+  ##
+  authUrlPath:
+  ## authUsername is used as username for authentication
+  ## of router
+  ##
+  authUsername: infracloud
+  ## jwtSigningKey is the signing key used for
+  ## signing the jwt token
+  ##
+  jwtSigningKey: converge2021
```
Comment (sanketsudake) [inline charts/fission-all/values.yaml:504]: Can we keep it `serverless` ?

File: charts/fission-all/values.yaml [hunk 8]
```diff
@@ -483,6 +482,37 @@ prometheus:
...
+  enabled: false
+  ## authUrlPath is the router path for
+  ## authentication endpoint, defaults to /auth/login
+  ## default '/auth/login'
+  ##
+  authUrlPath:
+  ## authUsername is used as username for authentication
+  ## of router
+  ##
+  authUsername: infracloud
```
Comment (sanketsudake) [inline charts/fission-all/values.yaml:500]: Please change user to `admin`.

File: go.mod [hunk 9]
```diff
@@ -10,6 +10,7 @@ require (
 	github.com/blend/go-sdk v1.20211025.3 // indirect
 	github.com/containerd/continuity v0.2.1 // indirect
 	github.com/dchest/uniuri v0.0.0-20200228104902-7aecb25e1fe5
+	github.com/dgrijalva/jwt-go v3.2.0+incompatible
```
Comment (sanketsudake) [inline go.mod:13]: This is deprecated library. Please use https://github.com/golang-jwt/jwt

File: go.sum [hunk 10]
```diff
@@ -258,6 +258,7 @@ github.com/dchest/uniuri v0.0.0-20200228104902-7aecb25e1fe5 h1:RAV05c0xOkJ3dZGS0
 github.com/dchest/uniuri v0.0.0-20200228104902-7aecb25e1fe5/go.mod h1:GgB8SF9nRG+GqaDtLcwJZsQFhcogVCJ79j4EdT0c2V4=
 github.com/deepmap/oapi-codegen v1.6.0/go.mod h1:ryDa9AgbELGeB+YEXE1dR53yAjHwFvE9iAUlWl9Al3M=
 github.com/denisenkom/go-mssqldb v0.10.0/go.mod h1:xbL0rPBG9cCiLr28tMa8zpbdarY27NDyej4t/EjAShU=
+github.com/dgrijalva/jwt-go v3.2.0+incompatible h1:7qlOGliEKZXTDg6OTjfoBKDXWrumCAMpl/TFQ4/5kLM=
```
Comment (sanketsudake) [inline go.sum:261]: This is deprecated library. Please use https://github.com/golang-jwt/jwt

**Review by sanketsudake — CHANGES_REQUESTED** [review]

File: go.sum (hunk 10 above)
Comment (praddy26) [reply go.sum]: Fixed

File: go.mod (hunk 9 above)
Comment (praddy26) [reply go.mod]: Fixed

File: charts/fission-all/values.yaml (hunk 8 above)
Comment (praddy26) [reply charts/fission-all/values.yaml]: Fixed

File: charts/fission-all/values.yaml (hunk 7 above)
Comment (praddy26) [reply charts/fission-all/values.yaml]: Fixed

File: charts/fission-all/values.yaml [hunk 11]
```diff
@@ -483,6 +482,37 @@ prometheus:
 canaryDeployment:
   enabled: false
 
+
```
Comment (sanketsudake) [inline charts/fission-all/values.yaml:485]: Can we add documentation for section above `authentication:` ?

Configuration options to enable authentication for fission function invocation via Fission router

File: pkg/fission-cli/cmd/token/command.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/spf13/cobra"
+
+	wrapper "github.com/fission/fission/pkg/fission-cli/cliwrapper/driver/cobra"
+	"github.com/fission/fission/pkg/fission-cli/flag"
+)
+
+func Commands() *cobra.Command {
+	createCmd := &cobra.Command{
+		Use:   "create",
+		Short: "Create a JWT token",
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/token/command.go:29]: Create a JWT token for fission function invocation via Fission router

Note: Authentication should be enabled to use this feature.

**Review by sanketsudake — CHANGES_REQUESTED** [review]

File: pkg/fission-cli/cmd/token/command.go (hunk 12 above)
Comment (praddy26) [reply pkg/fission-cli/cmd/token/command.go]: Fixed

File: charts/fission-all/values.yaml (hunk 11 above)
Comment (praddy26) [reply charts/fission-all/values.yaml]: Fixed

File: charts/fission-all/values.yaml [hunk 13]
```diff
@@ -483,6 +482,38 @@ prometheus:
...
+  ## in seconds
+  ## default '120'
+  ##
+  jwtExpiryTime: 
+  ## jwtIssuer is the issuer of JWT
+  ## default 'fission'
+  ##
+  jwtIssuer: fission
+
+
```
Comment (sanketsudake) [inline charts/fission-all/values.yaml:516]: We should have a single blank line.

File: charts/fission-all/values.yaml [hunk 14]
```diff
@@ -483,6 +482,38 @@ prometheus:
 canaryDeployment:
   enabled: false
 
+## Enable authentication for fission function invocation via Fission router
+##
+authentication:
```
Comment (sanketsudake) [inline charts/fission-all/values.yaml:487]: Please add relevant command to retrieve username and password if authencation is enabled. 

charts/fission-all/templates/NOTES.txt

**Review by sanketsudake — CHANGES_REQUESTED** [review]

File: charts/fission-all/values.yaml (hunk 13 above)
Comment (praddy26) [reply charts/fission-all/values.yaml]: Fixed

File: charts/fission-all/values.yaml (hunk 14 above)
Comment (praddy26) [reply charts/fission-all/values.yaml:487]: Fixed

**Review by sanketsudake — APPROVED** [review]

**Pull request merged**